### PR TITLE
Eric/training loss proposal

### PR DESCRIPTION
**Pull request by condnsdmatters**
Proposal for training loss interface.

### CODE REVIEW

File: gpflow/models/training_interface.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+    def training_loss_closure(self, data: Data, jit=True) -> Callable[[], tf.Tensor]:
+        training_loss = self.training_loss
+        if jit:
+            training_loss = tf.function(
+                self.training_loss
+            )  # TODO need to add correct input_signature here to allow for differently sized minibatches
+
+        def closure():
+            batch = next(data) if isinstance(data, collections.abc.Iterator) else data
+            return training_loss(data)
```
Comment (vdutor) [inline gpflow/models/training_interface.py]: This looks like a typo?
```suggestion
            return training_loss(batch)
```

File: gpflow/models/training_interface.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+        Specify a training loss for this model
+        """
+        raise NotImplementedError
+
+    def training_loss_closure(self, data: Data, jit=True) -> Callable[[], tf.Tensor]:
+        training_loss = self.training_loss
+        if jit:
+            training_loss = tf.function(
+                self.training_loss
+            )  # TODO need to add correct input_signature here to allow for differently sized minibatches
```
Comment (vdutor) [inline gpflow/models/training_interface.py]: how would you do this? By having a user specify `function_kwargs` as argument?

File: doc/source/notebooks/tailor/mixture_density_network.pct.py [hunk 3]
```diff
@@ -230,7 +230,7 @@ def moon_data(N, noise):
 
 # %%
 Scipy().minimize(
-    tf.function(model.training_loss_closure(data)),
+    model.training_loss_closure(data),
```
Comment (st--) [inline doc/source/notebooks/tailor/mixture_density_network.pct.py]:
```suggestion
    model.training_loss_closure(data, jit=True),
```
if you want to be consistent

File: gpflow/models/model.py [hunk 4]
```diff
@@ -56,24 +61,14 @@ def log_posterior_density(self, *args, **kwargs) -> tf.Tensor:
         """
         return self.maximum_likelihood_objective(*args, **kwargs) + self.log_prior_density()
 
-    def _training_loss(self, *args, **kwargs) -> tf.Tensor:
+    def maximum_a_posteriori_objective(self, *args, **kwargs) -> tf.Tensor:
```
Comment (st--) [inline gpflow/models/model.py]: you'd have to flip the sign as well

File: gpflow/models/model.py [hunk 5]
```diff
@@ -283,11 +239,31 @@ def predict_y(
...
+
+
+class InternalDataGPModel(GPModel, InternalDataTrainingInterface):
+    """
+    Base class for models that encapsulate their data and store it as
+    self.data; training_loss does not take any arguments.
+    """
+
+    def training_loss(self):
+        return self.maximum_a_posteriori_objective()
```
Comment (st--) [inline gpflow/models/model.py]: and flip the sign here as well as in the other one below, which I still think is a bit messy :/

Comment (st--) [reply gpflow/models/model.py]: I think I'd prefer to keep the `_training_loss()` and `_training_loss(data)`...

File: gpflow/models/model.py [hunk 6]
```diff
@@ -85,53 +80,14 @@ def maximum_likelihood_objective(self, *args, **kwargs) -> tf.Tensor:
...
-        if jit:
-            training_loss = tf.function(self.training_loss)  # TODO need to add correct input_signature here to allow for differently sized minibatches
-        else:
-            training_loss = self.training_loss
-
-        def minibatch_training_loss_closure():
-            batch = next(batch_iterator)
-            return training_loss(batch)
-
-        return minibatch_training_loss_closure
```
Comment (st--) [inline gpflow/models/model.py]: what about this?

File: gpflow/models/gpmc.py [hunk 7]
```diff
@@ -24,10 +24,10 @@
 from ..likelihoods import Likelihood
 from ..mean_functions import MeanFunction
 from ..utilities import to_default_float
-from .model import RegressionData, GPModel, MeanAndVariance, BayesianModelStoringData
+from .model import RegressionData, MeanAndVariance, InternalDataGPModel
 
 
-class GPMC(GPModel, BayesianModelStoringData):
+class GPMC(InternalDataGPModel):
```
Comment (st--) [inline gpflow/models/gpmc.py]: Note that for GPMC (and SGPMC in sgpmc.py) I explicitly re-defined training_loss - not because the code is different but because it's conceptually different and I wanted to highlight that... don't know how that interacts with your changes here?

File: gpflow/models/training_interface.py [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+        Specify a training loss for this model
+        """
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def training_loss_closure(self, *args, **kwargs):
+        """
+        Return a closure with a training loss for this module
+        """
+        raise NotImplementedError
```
Comment (joelberkeley-pio) [inline gpflow/models/training_interface.py]: I've just discussed with eric on slack about APIs with `(*args, **kwargs)`. They present a problem in that subclasses must also accept any number of arguments and keyword arguments to satisfy the contract. In particular, they have to have the signature `(*args: Any, **kwargs: Any)`.

You can think of it like this: Imagine you're working in a function which has signature
```python3
def foo(obj: TrainingInterface) -> None:
```
how do you work with it? I should be able to pass in whatever I want, for example
```
obj.training_loss(4, np.random, values='which', make=not 'sense')
```
This would fail if the runtime object doesn't support those arguments

Comment (condnsdmatters) [reply gpflow/models/training_interface.py]: do you mean the last sentence?  TBH am happy to get rid of the base `TrainingInterface`, to abide by liskov principle.

File: gpflow/models/gpmc.py (hunk 7 above)
Comment (condnsdmatters) [reply gpflow/models/gpmc.py]: If you reimplement in the derived it wont interact badly, though i dont know whether we can call the density a `loss`. Its very possible having `training loss` with `GPMC` and `SGPMC` makes no sense, and they should just inherit from `GPModel`

File: gpflow/models/model.py (hunk 5 above)
Comment (condnsdmatters) [reply gpflow/models/model.py]: i think this is much better. i think have a method which is _training_loss is much messier.

Comment (condnsdmatters) [inline gpflow/models/model.py]:
```suggestion
        return - self.maximum_a_posteriori_objective()
```

File: gpflow/models/model.py [hunk 9]
```diff
@@ -283,11 +239,31 @@ def predict_y(
...
+
+
+class ExternalDataGPModel(GPModel, ExternalDataTrainingInterface):
+    """
+    Base class for GP models that do not encapsulate the data; training_loss takes
+    data as argument.
+    """
+
+    def training_loss(self, data):
+        return self.maximum_a_posteriori_objective(data)
```
Comment (condnsdmatters) [inline gpflow/models/model.py]:
```suggestion
        return - self.maximum_a_posteriori_objective(data)
```

File: gpflow/models/training_interface.py (hunk 2 above)
Comment (condnsdmatters) [reply gpflow/models/training_interface.py]: i think you just specifiy the leading dim size to be None

File: gpflow/models/gpmc.py (hunk 7 above)
Comment (st--) [reply gpflow/models/gpmc.py]: we don't call the density a loss, but you might still want to do some gradient optimization first to relax the model before starting MCMC - that's why I chose to overwrite it

File: gpflow/models/training_interface.py (hunk 8 above)
Comment (st--) [reply gpflow/models/training_interface.py]: Aaaand it feels like we're back to square one. I would very much like to go with the thing that actually works in practice, for users, rather than the thing that makes the type system happy but makes it more cumbersome for users...

File: gpflow/models/training_interface.py (hunk 2 above)
Comment (st--) [reply gpflow/models/training_interface.py]: specify `input_signatures` argument to the tf.function - just have to look up the right TensorSpec ...

File: gpflow/models/training_interface.py [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+
+
+class ExternalDataTrainingInterface(TrainingInterface):
+    def training_loss(self, data):
+        """
+        Specify a training loss for this model
+        """
+        raise NotImplementedError
+
+    def training_loss_closure(self, data: Data, jit=True) -> Callable[[], tf.Tensor]:
```
Comment (st--) [inline gpflow/models/training_interface.py]: type signature doesn't agree with the use in the closure() below...?
```suggestion
    def training_loss_closure(self, data: Union[Data, Iterator], jit=True) -> Callable[[], tf.Tensor]:
```

File: gpflow/models/training_interface.py (hunk 2 above)
Comment (st--) [reply gpflow/models/training_interface.py]: this won't work in a general way for models that have different shapes of data -- e.g. RegressionData (SVGP) with data=(X,Y) vs InputData with data=Y (if there was a minibatching GPLVM) need different (incompatible) input_signatures.

File: gpflow/models/training_interface.py (hunk 8 above)
Comment (joelberkeley-pio) [reply gpflow/models/training_interface.py]: @st-- type systems aren't just there for theoretical aesthetics. I would guess that breaking the type system always results in at least one test that you can write that fails. It translates into code that's buggy in some way, and bugs are cumbersome for users

File: gpflow/models/gpmc.py [hunk 11]
```diff
@@ -24,10 +24,10 @@
 from ..likelihoods import Likelihood
 from ..mean_functions import MeanFunction
 from ..utilities import to_default_float
-from .model import RegressionData, MeanAndVariance, InternalDataGPModel
+from .model import RegressionData, MeanAndVariance, GPModel
 
 
-class GPMC(InternalDataGPModel):
+class GPMC(GPModel):
```
Comment (st--) [inline gpflow/models/gpmc.py]: Should this not remain an `InternalDataGPModel`?

File: gpflow/models/gpmc.py [hunk 12]
```diff
@@ -62,9 +62,6 @@ def __init__(
             loc=to_default_float(0.0), scale=to_default_float(1.0)
         )
 
-    def training_loss(self) -> tf.Tensor:
-        return -self.log_posterior_density()
-
```
Comment (st--) [inline gpflow/models/gpmc.py]: why did this get removed?

File: gpflow/models/training_interface.py (hunk 8 above)
Comment (st--) [reply gpflow/models/training_interface.py]: Python isn't a strongly-typed language, so from a `*args, **kwargs` signature you can't infer that you should be *allowed* to pass in whatever you want.

File: gpflow/models/model.py [hunk 13]
```diff
@@ -56,23 +61,13 @@ def log_posterior_density(self, *args, **kwargs) -> tf.Tensor:
         """
         return self.maximum_likelihood_objective(*args, **kwargs) + self.log_prior_density()
 
-    def _training_loss(self, *args, **kwargs) -> tf.Tensor:
+    def maximum_a_posteriori_objective(self, *args, **kwargs) -> tf.Tensor:
         """
         Minimization objective for TensorFlow optimizers (including
         gpflow.optimizers.Scipy). Includes the log prior density for maximum
         a-posteriori (MAP) estimation.
```
Comment (st--) [inline gpflow/models/model.py]: Docstring inconsistent with function name change

File: gpflow/models/training_interface.py (hunk 8 above)
Comment (st--) [reply gpflow/models/training_interface.py]: Trying to force static typing on Python can result in unpythonic and cumbersome code. I don't think that's the right approach for a research toolbox that's supposed to be flexible and make it easy for GP researchers to implement their ideas.

Comment (joelberkeley-pio) [reply gpflow/models/training_interface.py]: > Python isn't a strongly-typed language, so from a `*args, **kwargs` signature you can't infer that you should be _allowed_ to pass in whatever you want.

it's not a matter of it being statically typed (python is strongly typed, but is optionally statically typed), it's a fundamental design principle of OOP (LSP), which applies equally for dynamic and static typing

> 

Trying to force static typing on Python can result in unpythonic and cumbersome code. I don't think that's the right approach for a research toolbox that's supposed to be flexible and make it easy for GP researchers to implement their ideas.

I don't agree it's unpythonic. Static types are after all part of the standard library. Cumbersome? Yes, if overused. But overused and cumbersome are both subjective. Regarding a research toolbox, I'm not really clear yet on what that means. I hope time will tell. Flexible? I don't anticipate this will make it any more or less flexible.

Comment (joelberkeley-pio) [reply gpflow/models/training_interface.py]: I suspect the concern here is the conceptual overhead of the design. I agree it's a headache and not something we should keep in the code base. I guess the discussion is what to do until we reach that point. I lean on the side of dealing with it safely until we can converge on a final design. I appreciate that's not the universal opinion.

File: gpflow/models/model.py [hunk 14]
```diff
@@ -85,53 +80,14 @@ def maximum_likelihood_objective(self, *args, **kwargs) -> tf.Tensor:
...
-
-    def training_loss_closure(self, jit=True) -> Callable[[], tf.Tensor]:
-        if jit:
-            return tf.function(self.training_loss)
-        else:
-            return self.training_loss
-
-
-class BayesianModelNotStoringData(BayesianModel):
+class ExternalDataBayesianModel(BayesianModel, ExternalDataTrainingInterface):
```
Comment (joelberkeley-pio) [inline gpflow/models/model.py]: this appears only to be used in the notebooks. do we want to have something in the codebase that's only there for a single notebook?

File: gpflow/models/model.py [hunk 15]
```diff
@@ -283,11 +239,31 @@ def predict_y(
...
+
+
+class ExternalDataGPModel(GPModel, ExternalDataTrainingInterface):
+    """
+    Base class for GP models that do not encapsulate the data; training_loss takes
+    data as argument.
+    """
+
+    def training_loss(self, data):
+        return -self.maximum_a_posteriori_objective(data)
```
Comment (joelberkeley-pio) [inline gpflow/models/model.py]: you could completely decouple these from the `GPModel`, which may help with keeping the hierarchy simple, using
```python3
class InternalDataGPModel(GPModel, InternalDataTrainingInterface):
    @abstractmethod
    def maximum_a_posteriori_objective(self):
        pass

    def training_loss(self):
        return -self.maximum_a_posteriori_objective()
```

File: gpflow/models/model.py [hunk 16]
```diff
@@ -56,23 +61,13 @@ def log_posterior_density(self, *args, **kwargs) -> tf.Tensor:
         """
         return self.maximum_likelihood_objective(*args, **kwargs) + self.log_prior_density()
 
-    def _training_loss(self, *args, **kwargs) -> tf.Tensor:
+    def maximum_a_posteriori_objective(self, *args, **kwargs) -> tf.Tensor:
```
Comment (joelberkeley-pio) [inline gpflow/models/model.py]: i feel the same about `(*args, **kwargs)` here

File: gpflow/models/model.py (hunk 14 above)
Comment (joelberkeley-pio) [reply gpflow/models/model.py]: do we expect users to use it?

File: gpflow/models/model.py (hunk 15 above)
Comment (joelberkeley-pio) [reply gpflow/models/model.py]: this would also make the mixin work with `GPModel` and `BayesianModel` without assuming either

Comment (joelberkeley-pio) [reply gpflow/models/model.py]: they could even be merged with `InternalDataTrainingInterface` and `ExternalDataTrainingInterface` then

File: gpflow/models/training_interface.py [hunk 17]
```diff
@@ -0,0 +1,46 @@
...
+import numpy as np
+
+InputData = tf.Tensor
+OutputData = tf.Tensor
+RegressionData = Tuple[InputData, OutputData]
+Data = TypeVar("Data", RegressionData, InputData)
+
+
+class InternalDataTrainingInterface:
+    def training_loss(self):
```
Comment (joelberkeley-pio) [inline gpflow/models/training_interface.py]:
```suggestion
class InternalDataTrainingInterface(ABC):
    @abstractmethod
    def training_loss(self):
```
etc?

File: gpflow/models/model.py (hunk 15 above)
Comment (st--) [reply gpflow/models/model.py]: The training_loss / maximum_a_posteriori_objective should go with BayesianModel not GPModel.

File: gpflow/models/model.py (hunk 14 above)
Comment (st--) [reply gpflow/models/model.py]: yes.

File: gpflow/models/sgpmc.py [hunk 18]
```diff
@@ -86,9 +86,6 @@ def __init__(
             loc=to_default_float(0.0), scale=to_default_float(1.0)
         )
 
-    def training_loss(self) -> tf.Tensor:
-        return -self.log_posterior_density()
-
```
Comment (st--) [inline gpflow/models/sgpmc.py]: can you put this back?

File: gpflow/models/sgpmc.py [hunk 19]
```diff
@@ -24,11 +24,11 @@
 from ..likelihoods import Likelihood
 from ..mean_functions import MeanFunction
 from ..utilities import to_default_float
-from .model import RegressionData, GPModel, MeanAndVariance, BayesianModelStoringData
+from .model import RegressionData, MeanAndVariance, GPModel
 from .util import inducingpoint_wrapper
 
 
-class SGPMC(GPModel, BayesianModelStoringData):
+class SGPMC(GPModel):
```
Comment (st--) [inline gpflow/models/sgpmc.py]: can you put the training interface back?

File: gpflow/models/model.py (hunk 15 above)
Comment (joelberkeley-pio) [reply gpflow/models/model.py]: ok. that's fine. it's just that we have the option to do either

**Comment by joelberkeley-pio** [comment]
I'd like to see this PR compared to develop, but I don't want to edit your PR. When can we see that?

**Comment by st--** [comment]
> I'd like to see this PR compared to develop, but I don't want to edit your PR. When can we see that?

How about `git diff origin/develop origin/eric/training-loss-proposal gpflow` ?

File: gpflow/models/model.py (hunk 15 above)
Comment (condnsdmatters) [reply gpflow/models/model.py]: im not entirely sure what that buys you, except that once again ppl need to remember to add the mixin for their new derived model (which im not so keen on). lets discuss over zoom

**Comment by st--** [comment]
Closed in favour of #1362
